src/edu/csus/ecs/pc2/shadow/RemoteEventFeedMonitor.java

```diff
@@ -251,504 +259,445 @@ public void run() {
+                                           //  of the event stream which specify a "create" op and have "null" for values such as "judgement_type_id".
 
-                                        } catch (Exception e) {
-                                            // TODO design error handling reporting (logging?)
-                                            logAndDebugPrint(log, Level.SEVERE, "Exception processing event: " + event, e);
+                                        // this (appears to be) a judgement we want; save it in the global judgements map under a key of
+                                        // the judgement ID with value "submissionID:judgement"
+//                                                  System.out.println ("Adding judgement " + judgementID + " for submission " + submissionID + " with judgement " + judgement + " to RemoteJudgements Map");
+                                        synchronized (remoteJudgementsMapLock) {
+                                            getRemoteJudgementsMap().put(judgementID, submissionID + ":" + judgement);
                                         }
+                                    }
+                                }
+
+                            } catch (Exception e) {
+                                // TODO design error handling reporting (logging?)
+                                logAndDebugPrint(log, Level.SEVERE, "Exception processing event: " + dataMap.toString(), e);
+                            }
 
-                                    } else if ("state".equals(eventType)) {
+                        } else if ("state".equals(notifyType)) {
 
-                                        // State messages look like this (Sample from DOMjudge Event feed Dhaka WF 2022):
-                                        // {"token":"151701","id":null,"type":"state",
-                                        //  "data":{"started":"2022-11-10T10:53:36.000+06:00","ended":null,"frozen":null,"thawed":null,"finalized":null,"end_of_updates":null},"time":"2022-11-10T10:53:36.014+06:00"}
-                                        // {"token":"229483","id":null,"type":"state",
-                                        //  "data":{"started":"2022-11-10T10:53:36.000+06:00","ended":null,"frozen":"2022-11-10T14:53:36.000+06:00","thawed":null,"finalized":null,"end_of_updates":null},"time":"2022-11-10T14:53:36.112+06:00"}
-                                        // {"token":"257177","id":null,"type":"state",
-                                        //  "data":{"started":"2022-11-10T10:53:36.000+06:00","ended":"2022-11-10T15:53:36.000+06:00","frozen":"2022-11-10T14:53:36.000+06:00","thawed":null,"finalized":null,"end_of_updates":null},"time":"2022-11-10T15:53:36.035+06:00"}
+                            // State messages look like this (Sample from DOMjudge Event feed Dhaka WF 2022):
```

Comment:
It worries me that this comment (and the code following it) seem to assume that the format of State messages remains the same regardless of which API version the Remote CCS is using.  That seems patently incorrect -- which in turn would seem to imply that we can only use this class if the Remote CCS is using a particular CLICS API (which in this case seems to be 2020-03?)  

Is it the case that the following code handles the differences between (known) API versions?

Reply:
Yes.  The `RemoteEventFeedMonitor `figures out the format on-the-fly based on what's in the notification part of the message.  For the moment, this works for 2020-03, 2022-07 and 2023-06.   I have to create an issue after all this merged indicating what has to be moved into the "per protocol" packages.  This class is quite large and the guts of it will be the same for all protocols.  What we need is an interface that can extract the needed information from whatever format feed is being read, and `RemoteEventFeedMonitor `can use that interface.  Some reports also generate JSON, but only generate it for 2023-06; there should be some chooser in the report generator that allows you to pick the protocol version to generate the JSON for. (I have this in my list as well).